selftests/unit/test_plugin_vmimage.py

```diff
@@ -0,0 +1,100 @@
+        config_file.write(temp_settings)
+        config_file.close()
+        return base_dir, mapping, config_file.name
+
+    def _create_test_files(self):
+        with unittest.mock.patch('avocado.core.data_dir.settings.settings', self.stg):
+            expected_images = [{'name': 'CentOS', 'file': 'CentOS-{version}-{arch}-GenericCloud-{build}.qcow2.xz$'},
+                               {'name': 'Debian', 'file': 'debian-{version}-openstack-{arch}.qcow2$'},
+                               {'name': 'JeOS', 'file': 'jeos-{version}-{arch}.qcow2.xz$'},
+                               {'name': 'OpenSUSE', 'file': 'openSUSE-Leap-{version}-OpenStack.{arch}-{build}.qcow2$'},
+                               {'name': 'Ubuntu', 'file': 'ubuntu-{version}-server-cloudimg-{arch}.img'}
+                               ]
+            cache_dir = data_dir.get_cache_dirs()[0]
+            providers = [provider() for provider in vmiage_util.list_providers()]
+
+            for provider in providers:
+                for image in expected_images:
+                    if image['name'] == provider.name:
+                        path = os.path.join(cache_dir, "vmimage", image['name'],
+                                            str(provider.version), provider.arch)
+                        os.makedirs(path)
+                        image['version'] = str(provider.version)
+                        image['arch'] = provider.arch
+                        image['build'] = "1234"
+                        image['file'] = os.path.join(path, image['file'].format(
+                            version=provider.version,
+                            build=image['build'],
+                            arch=provider.arch).replace('$', ''))
+                        open(image["file"], "a").close()
+            return sorted(expected_images, key=lambda i: i['name'])
+
+    def setUp(self):
+        (self.base_dir, self.mapping,
+         self.config_file_path) = self._get_temporary_dirs_mapping_and_config()
+        self.stg = settings.Settings(self.config_file_path)
+        self.expected_images = self._create_test_files()
+
+    def test_list_downloaded_images(self):
+        with unittest.mock.patch('avocado.core.data_dir.settings.settings', self.stg):
+            images = sorted(vmimage_plugin.list_downloaded_images(), key=lambda i: i['name'])
+            for index, image in enumerate(images):
+                for key in image:
+                    self.assertEqual(self.expected_images[index][key], image[key],
+                                     "Founded image is different from the expected one")
+
+    @unittest.skipIf(int(os.environ.get("AVOCADO_CHECK_LEVEL", 0)) < 2,
+                     "Skipping test that take a long time to run, are "
+                     "resource intensive or time sensitve")
+    @unittest.skipIf(missing_binary('qemu-img'),
+                     "QEMU disk image utility is required by the vmimage utility ")
+    def test_download_image(self):
```

Comment:
I like the test, but don't like the 300+ MB download... I've always wanted to write a CirrOS image provider, so this gave me the motivation to do it, so we can have a 15 MB download instead.  See:

https://github.com/avocado-framework/avocado/pull/3286

Also, consider checking that we have network access, and if not, skip the test.  See `selftests/functional/test_thirdparty_bugs.py` for an example.

Reply:
Thanks for the PR I will use it.